209 add stylized fonts

wreckx opened the pull request:
The font could be better but the important part is that i modified the ui so that all we need to do now is just add the desired font and change all the fonts to the one you want with minimal effects on other parts of the ui unlike the former where changing the font on one part changes the fonts on other parts you dont want.

alrauscher commented:
I like this idea, but I agree the font is not the best. Also, three tests are failing. It looks like the files for the fonts just need to be mocked out. I'm going to modify the tests to account for the fonts and commit to the PR.

alrauscher commented on src/userinterface.py:
```diff
@@ -22,13 +22,13 @@ class UserInterface:
-        self.font_fixed_large: pygame.font = pygame.font.SysFont("Courier", 35, True)
-        self.font_fixed_xlarge: pygame.font = pygame.font.SysFont("Courier", 60, True)
+        self.font_fixed_large: pygame.font = pygame.font.Font('assets/font/slkscreb.ttf', 35)
+        self.font_fixed_xlarge: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 60)
```
I feel like the fonts need to be setup in the assets. It will make it so much easier to swap out.

jcooke-dev commented on src/userinterface.py:
```diff
@@ -21,14 +21,37 @@ class UserInterface:
+        # Font setups
+        # font - logo (splash screen fonts)
+        self.font_logo: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 60)
+        self.font_logo_tagline: pygame.font = pygame.font.Font('assets/font/slkscre.ttf', 24)
+        # font - main menu
+        self.menu_main_font: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 72)
+        self.menu_sub_font: pygame.font = pygame.font.Font('assets/font/slkscr.ttf', 32)
+        # font - main menu elements
+        self.h2p_font: pygame.font = pygame.font.Font('assets/font/slkscr.ttf', 30)
+        self.font_credits: pygame.font = pygame.font.Font('assets/font/slkscre.ttf', 40)
+        self.font_leaderboard: pygame.font = pygame.font.Font('assets/font/slkscreb.ttf', 35)
+        self.back_btn_font: pygame.font = pygame.font.Font('assets/font/slkscr.ttf', 36)
+        #font - playing menus (game_over, pause, and highscore)
+        self.font_title_text: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 70)
+        self.font_subtitle_text: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 60)  # high scores
+        self.font_buttons: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 40)
+        # font - play screen elements
+        self.game_intro_font: pygame.font = pygame.font.Font('assets/font/slkscreb.ttf', 40)
+        self.status_font: pygame.font = pygame.font.Font('assets/font/slkscrb.ttf', 36)
         # not certain this will reliably get a font (especially on diff OSes), but it's supposed to
         # fall back to a default pygame font
-        self.font_fixed_small: pygame.font = pygame.font.SysFont("Courier", 16, True)
-        self.font_fixed_large: pygame.font = pygame.font.SysFont("Courier", 35, True)
-        self.font_fixed_xlarge: pygame.font = pygame.font.SysFont("Courier", 60, True)
+        self.font_fixed_small: pygame.font = pygame.font.SysFont("Courier", 16, True)  # dev overlay font
+
```
I'd put the actual font creation over in assets, similar to how it's been done with art and music (so the FILENAME definitions and then actual font creation in load_assets()).
example:
FONT_DIR = 'assets/font/'
FONT_SLKSCR_FILENAME = 'slkscr.ttf'
...
FONT_SLKSCR_30 = pygame.font.Font(os.path.join(FONT_DIR, FONT_SLKSCR_FILENAME), 30)

You could still assign whichever fonts make sense to these UI.self vars in here (from assets).  That would keep any redundancy to a minimum, but allow you to map the separate fonts to the custom UI elements as you're already doing.

jcooke-dev commented:
So, as Ann noted, I'd stick the font paths/filenames and creation over in the assets.py, just for consistency with how we're already doing it with art and music.  It still makes sense to assign whichever fonts you want to use to their respective UI elements in the UI.__init__.

Also, my personal take is that this is not the font we want to use (and I know you were mainly refactoring the font creation and assignment to make it more flexible and not choosing a final font family :) ).  There are situations where a fixed-width font is best (not just in the overlay), such as with the Leaderboard (and character alignment).  Maybe we could go back to using the fonts already in place?  That could still take advantage of your code restructuring -- those fonts would just be added to the creation over in assets.py and assigned to their elements in UI.

jcooke-dev commented on src/userinterface.py:
```diff
@@ -77,17 +100,23 @@ def draw_pause_menu(self) -> tuple[pygame.Rect, pygame.Rect, pygame.Rect]:
+        self.surface.blit(title1_text, title1_text_rect)
+        self.surface.blit(title2_text, title2_text_rect)
+
+        button_width = 300
```
To keep going with your font flexibility approach, you may consider turning these types of dimensions (throughout the UI) into dynamically created off of the font rect (in draw_button()?).

wreckx commented:
Since we arent gonna be using that particular font anyways, i just removed it from the branch. rather than adding stylized fonts, i ended up just refactoring the userinterface.py to be cleaner and more readable. i refactored the draw_button to add to its functionality and used that to recreate all the buttons in the main menu.

jcooke-dev commented:
Looks good -- thanks for cleaning up UI (and making much of the code more consistent)!
